src/Donors/DonorsQuery.php

```diff
@@ -0,0 +1,129 @@
+<?php
+
+namespace Give\Donors;
+
+use Give\Donations\ValueObjects\DonationMetaKeys;
+use Give\Donors\Models\Donor;
+use Give\Framework\Models\ModelQueryBuilder;
+use Give\Framework\QueryBuilder\JoinQueryBuilder;
+use Give\Framework\QueryBuilder\QueryBuilder;
+
+/**
+ * @unreleased
+ */
+class DonorsQuery
+{
+    /**
+     * @var ModelQueryBuilder<Donor>
+     */
+    protected $query;
+
+    /**
+     * @unreleased
+     */
+    public function __construct()
+    {
+        $this->query = Donor::query();
+    }
+
+    /**
+     * Delegates methods not defined locally to the underlying query.
+     *
+     * @unreleased
+     *
+     * @return mixed
+     */
+    public function __call(string $method, array $args)
+    {
+        if (method_exists($this, $method)) {
+            return $this->$method(...$args);
+        }
+
+        return $this->query->$method(...$args);
+    }
+
+    /**
+     * @unreleased
+     */
+    public function whereDonorsHaveDonations(
+        string $mode = 'live',
+        int $campaignId = 0,
+        bool $excludeAnonymousDonors = true
+    ): self {
+        $this->query->join(function (JoinQueryBuilder $builder) use ($mode) {
+            // The donationmeta1.donation_id should be used in other "donationmeta" joins to make sure we are retrieving data from the proper donation
+            $builder->innerJoin('give_donationmeta', 'donationmeta1')
+                ->joinRaw("ON donationmeta1.meta_key = '" . DonationMetaKeys::DONOR_ID . "' AND donationmeta1.meta_value = ID");
+
+            // Include only current payment "mode"
+            $builder->innerJoin('give_donationmeta', 'donationmeta2')
+                ->joinRaw("ON donationmeta2.meta_key = '" . DonationMetaKeys::MODE . "' AND donationmeta2.meta_value = '{$mode}' AND donationmeta2.donation_id = donationmeta1.donation_id");
+        });
+
+        if ($campaignId) {
+            // Filter by CampaignId - Donors only can be filtered by campaignId if they donated to a campaign
+            $this->query->join(function (JoinQueryBuilder $builder) use ($campaignId) {
+                $builder->innerJoin('give_donationmeta', 'donationmeta3')
+                    ->joinRaw("ON donationmeta3.meta_key = '" . DonationMetaKeys::CAMPAIGN_ID . "' AND donationmeta3.meta_value = {$campaignId} AND donationmeta3.donation_id = donationmeta1.donation_id");
+            });
+        }
+
+        if ($excludeAnonymousDonors) {
+            // Exclude anonymous donors from results - Donors only can be excluded if they made an anonymous donation
+            $this->query->join(function (JoinQueryBuilder $builder) {
+                $builder->innerJoin('give_donationmeta', 'donationmeta4')
+                    ->joinRaw("ON donationmeta4.meta_key = '" . DonationMetaKeys::ANONYMOUS . "' AND donationmeta4.meta_value = 0 AND donationmeta4.donation_id = donationmeta1.donation_id");
```

Comment:
Hey Glauber, just a note here. You don't have to use `joinRaw`, you can use `on`, `andOn`, etc., check the docs https://github.com/impress-org/givewp/tree/develop/src/Framework/QueryBuilder#advanced-join-clauses

Reply:
I used the `joinRaw` previously because I had a problem with quotes, but I just noticed that we have a `$quote` parameter in the docs. Thanks for sharing! 

Resolved in 5b22aabea7e7250cbf1377064e1796a12fbd86d3